include/internal/tsan_assist.h

```diff
@@ -56,21 +58,57 @@
 #  define tsan_load(ptr) __atomic_load_n((ptr), __ATOMIC_RELAXED)
 #  define tsan_store(ptr, val) __atomic_store_n((ptr), (val), __ATOMIC_RELAXED)
 #  define tsan_counter(ptr) __atomic_fetch_add((ptr), 1, __ATOMIC_RELAXED)
+#  define tsan_ld_acq(ptr) __atomic_load_n((ptr), __ATOMIC_ACQUIRE)
+#  define tsan_st_rel(ptr, val) __atomic_store_n((ptr), (val), __ATOMIC_RELEASE)
 # endif
 
-#elif defined(_MSC_VER) && _MSC_VER>=1200
-
+#elif defined(_MSC_VER) && _MSC_VER>=1200 \
+      && (defined(_M_IX86) || defined(_M_AMD64) || defined(_M_X86) || \
+          defined(_M_ARM64) || (defined(_M_ARM) && _M_ARM >= 7))
+/*
+ * There is subtle dependency on /volatile:<iso|ms> command-line option.
+ * "ms" implies same semantic as memory_order_acquire for loads and
+ * memory_order_release for stores, while "iso" - memory_order_relaxed for
+ * either. Real complication is that defaults are different on x86 and ARM.
+ * There is explanation for that, "ms" is backward compatible with earlier
+ * compiler versions, while multi-processor ARM can be viewed as brand new
+ * platform to MSC and its users, and with non-relaxed semantic taking toll
+ * with additional instructions and penalties, it kind of makes sense to
+ * default to "iso"...
+ */
 # define TSAN_QUALIFIER volatile
-# define tsan_load(ptr) (*(ptr))
-# define tsan_store(ptr, val) (*(ptr) = (val))
+# if defined(_M_ARM) || defined(_M_ARM64)
+#  define _InterlockedExchangeAdd _InterlockedExchangeAdd_nf
+#  pragma intrinsic(_InterlockedExchangeAdd_nf)
+#  pragma intrinsic(__iso_volatile_load32, __iso_volatile_store32)
+#  ifdef _WIN64
+#   define _InterlockedExchangeAdd64 _InterlockedExchangeAdd64_nf
+#   pragma intrinsic(_InterlockedExchangeAdd64_nf)
+#   pragma intrinsic(__iso_volatile_load64, __iso_volatile_store64)
+#   define tsan_load(ptr) (sizeof(*(ptr)) == 8 ? __iso_volatile_load64(ptr) \
+                                               : __iso_volatile_load32(ptr))
+#   define tsan_store(ptr, val) (sizeof(*(ptr)) == 8 ? __iso_volatile_store64((ptr), (val)) \
+                                                     : __iso_volatile_store32((ptr), (val)))
```

Comment:
Since this is supposed to be relaxed, can't we just do *ptr, and *ptr = val?

Reply:
Which I guess is the old code ... I guess I must be missing something.

Reply:
Rationale here is that if you pass /volatile:ms as additional argument at config time, there will be fences at just pointer references. Yes, I've verified it. Note that this is _M_ARM[64] we are looking at here.

Reply:
Maybe it's an option not to use the volatile keyword in case _ISO_VOLATILE is defined?

Reply:
It's a relaxed write, we only care that the write happen atomically, we don't care when. It can't move it outside the lock/unlock. I fail to see the need for the volatile.

I really wish Microsoft would support newer C standards. I think at some point we should stop saying it's supported.

Reply:
> It can't move it outside lock/unlock. I fail to see the need for the volatile.

In my mind it's not about seeing the need here and now, but rather about not making assumption that regular reference would always do. I mean it's not about *this* case, but what can be useful in some other *arbitrary* case. Or in other words question is if guarantee that tsan_load/store are not optimized away completely can be useful, and the answer is "it *might* be". Which is arguably sufficient for keeping it.

Reply:
As far as I know, it will generate the same assembler for `*(ptr) = (val)` and `atomic_store_explicit((ptr), (val), memory_order_relaxed)` (on ARMv7). Relaxed is really just like the assignment. The only guarantee that the atomic_store_explicit() gives you over the assignment is that it will happen atomicly, using a lock if required. And it doesn't need a lock on ARMv7.

Reply:
That is, without the volatile qualifier should be enough.

Reply:
> without the volatile qualifier should be enough.

In **this** case, yes. Question is about **general** case. Again, volatile is different from just `*(ptr) = (val)` because it can't be optimized away (for example coalesced with `*(ptr) = (val2)`) or reordered with other volatile references.

Reply:
There seem to be some mix up here. These are comments to Windows-for-ARM-specific segment, yet atomic_store_explicit was mentioned. I mean there is no atomic_store_explicit on Windows (there is one in C++, but not C), so what exactly are we talking about here?

Reply:
I have a feeling we've been talking past each other for a while on this.

The other versions of tsan_load() and tsan_store() do not provide this guarantee that it will not be coalesced.

From Herb Sutter's slides on atomics vs volatile:

what | atomic | volatile
---- | ---------- | -----------
Reorder/invent/elide allowed, such as combining these special ops themselves? | Some optimizations are allowed, such as combining these special ops to the same location |  No optimization possible; the compiler is not allowed to assume it knows anything about the type... not even v = 1; r1 = v; -> v = 1; r1=1;

Reply:
For starters, as already established, Microsoft C doesn't have atomic type (C++ does, C doesn't), volatile is all we get. Then there is /volatile:ms which implies acquire semantic on volatile loads and release semantic on volatile stores. And there is /volatile:iso which implies relaxed semantic on volatile loads and stores. Again, volatile is *all* we get from MSC. And it comes with at least "never coalesce" guarantee. And I argue that it's appropriate to take it, the said guarantee, for tsan_load and tsan_store.

As for table. Since Herb is talking essentially about Microsoft compiler, the table ought to apply to C++, not C. Is it possible that C++ allows combining/coalescing of atomic operations, and C doesn't? At least C11 spec revolves very much around object visibility and modification order, and I don't quite see how would even coalescing be allowed. You even find note that reads "This effectively disallows compiler reordering of atomic operations to a single object, even if both operations are ‘‘relaxed’’ loads." Which I interpret as "relaxed" load is most permissive operation, yet if two appear in source code, there will be two operations in machine code. And since it's most permissive operation, less permissive ones will also translate to distinct machine code operations, instructions or instruction sequences.